records/2022-07-20.security.prefix-api-token.md

```diff
@@ -0,0 +1,39 @@
+# TITLE
+
+Date: 2022-07-20
+
+## Status
+
+**DRAFT**
+
+## Context
+
+The mishandling of application secrets has lead to many security incidents in companies big and small. Given the inherent risk in how users store these secrets, it is important that we provide mechanisms to protect our users from doing so improperly.  Taking a cue from [Github](https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/), [Slack](https://api.slack.com/authentication/token-types), and [Stripe](https://stripe.com/docs/api/authentication), we see that token prefixes are an effective technique to help us ensure our tokens don't end up in public repositories.
+
+By combining a known prefix with Github's [secret scanning program](https://docs.github.com/en/developers/overview/secret-scanning-partner-program), we can be alerted if one of our users mistakenly stores their API keys in the open and we can take action.
+
+## Options
+
+- Add a known prefix to our API keys in order to be able to easily identify them
+- Register with Github's secret scanning program and create an endpoint for them to report issues back to us
+- Give users a grace period to generate new keys
+  - After grace period, enforce a rule that ensures all API keys contain our prefix
+
+## Additional considerations
+
+### Supporting older keys
+We want to be able to support previous API keys for a certain time period and there should be nothing additional we need to do for that (provided we dont remove old keys). But, in the future, we may want to enforce that all keys have the prefix after.
+
+In order to make this work we would need to communicate with existing users who have older keys and then invalidate the keys once we know they have new ones.
+
+### Verifying API keys
+Currently, API ignores everything but the right-most characters of an API key.  This will need to be updated in order to verify the existence of the prefix.
+
+
```

Comment:
We'd need a section on the [API client libraries](https://documentation.notification.canada.ca/en/clients.html) to see if these would stay compatible with the new format of the keys.

Reply:
Good point; will add this.

Reply:
Added.